add mine flag in cli for updates & overrides query

ryanlerch opened the pull request:
Implement a --mine flag for `bodhi updates query` & `bodhi overrides query`. The flag uses the USERNAME envvar as the user parameter for the query so a user can quickly view their own updates or overrides.

Fixes #811 #1382

bowlofeggs commented:
test

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -43,6 +43,48 @@ def _warn_if_url_and_staging_set(ctx, param, value):
+        Boolean: The value of the --mine flag.
+    """
+    if value and not os.environ['USERNAME']:
+        raise click.BadParameter('USERNAME environment variable not set.')
```
I recently created an [```_init_username()```](https://github.com/fedora-infra/bodhi/blob/2.7.0/bodhi/client/bindings.py#L335-L363) method that can be used to get the user's username automatically. It uses their saved cookies to find if they've ever authenticated before and uses that name if they have. Otherwise, it falls back to prompting the user. The environment variable will override what this does since it will set username when the BodhiClient is initialized.

What if we make that a public method (by renaming it from ```_init_username``` to ```init_username```) and use that to ensure that the username will work as people expect? This way, they don't have to set an environment variable (though they still can).

ryanlerch replied:
Bit confused with this here. I like the init_username idea, but with the current check here in place, the validator won't pass, and we wont get down to running the init_username.

IMHO, we could just use the init_username for the --mine, and not worry about the environment variable at all. This means that a user that has logged in before, --mine will just work, otherwise, they will need to type in their username every time --mine is run.

Of course, once we enable openidconnect, this will make a little more sense i think.

bowlofeggs replied:
I think the validator is still OK - if the user supplies ```--username``` or the environment variable, those things override what ```init_username``` does (it'll just no-op if ```self.username``` is already set). So the validator checks are fine and I think The Right Thing™ will happen whether or not the user uses the flag or env var. ```init_username()``` only takes action when the user doesn't tell us who they are through those mechanisms.

ryanlerch replied:
I had the init_username in the main method for the query methods, so if the envvar was not set, it never got down to the init_username() call, becasue the validator would fail on not having an envvar set. going to move the init_username() call up to the validator, and call it only if the envvar is not set.

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -43,6 +43,48 @@ def _warn_if_url_and_staging_set(ctx, param, value):
+    if value and not os.environ['USERNAME']:
+        raise click.BadParameter('USERNAME environment variable not set.')
+    elif value and ctx.params.get('user'):
+        # also check in this callback, in case the --mine was given after --user
```
In your testing, does the order of the parameters affect whether each of the callbacks see the other setting?

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -299,9 +344,12 @@ def query(url, **kwargs):
     """
     client = bindings.BodhiClient(base_url=url, staging=kwargs['staging'])
+    if mine:
+        kwargs['user'] = os.environ['USERNAME']
```
If you put a call to ```client.init_username``` above this, I think you can just do ```kwargs['user'] = client.username``` here.

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -506,11 +557,14 @@ def query_buildroot_overrides(url, user=None, **kwargs):
     """
     client = bindings.BodhiClient(base_url=url, staging=kwargs['staging'])
+    if mine:
+        user = os.environ['USERNAME']
```
Similar recommendations here.

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -43,6 +43,48 @@ def _warn_if_url_and_staging_set(ctx, param, value):
+    if value and ctx.params.get('mine'):
+        click.echo('Warning: --mine and --user options are both set. --user will be ignored.\n',
+                   err=True)
+    return value
```
If both validators are really needed, and since this validator only prints a warning, you can probably save writing this code twice (since the other validator has the same check) and move this warning into the bodies of the functions that use them.

bowlofeggs commented:
On Mon, 2017-06-26 at 21:32 -0700, Ryan Lerch wrote:
> I had the init_username in the main method for the query methods, so
> if the envvar was not set, it never got down to the init_username()
> call, becasue the validator would fail on not having an envvar set.
> going to move the init_username() call up to the validator, and call
> it only if the envvar is not set.

Ah I think I see why it's confusing a little better now - I don't think
you need to validate that --user or USER is set when --mine is used. If
those aren't set, init_username() will prompt the user for a username.
So I suggest not worrying about validating and just let it prompt when
needed. Does that help?

ryanlerch commented:
Okies, updated the code. Just need to fix up the tests -- i started on it, but need to finish it off.

ryanlerch committed:
add mine flag in cli for updates & overrides query

Implement a --mine flag for `bodhi updates query` &
`bodhi overrides query`. The flag uses the USERNAME envvar
as the user parameter for the query so a user can quickly
view their own updates or overrides.

Fixes #811 #1382

ryanlerch commented:
Okies, fixed the tests, this one should be right for a re-review!

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -285,10 +285,12 @@ def edit(user, password, url, **kwargs):
               is_flag=True, default=False)
 @url_option
-def query(url, **kwargs):
+def query(url, mine=None, **kwargs):
```
Optional: For consistency with the type, you could make this default to ```False``` instead of ```None```. No big deal though, as it is functionally the same.

ryanlerch replied:
done

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -285,10 +285,12 @@ def edit(user, password, url, **kwargs):
               type=click.Choice(['newpackage', 'security', 'bugfix', 'enhancement']))
 @click.option('--user', help='Updates submitted by a specific user')
+@click.option('--mine', is_flag=True,
+              help='Show only your updates. User is defined by the $USERNAME envvar')
```
Technically, if the user has logged in before the user is also defined by their login cookies. And if they both don't set ```USERNAME``` and haven't logged in, ```init_username()``` will prompt them for a username. Thus, I think we should either just not mention the ```USERNAME``` bit, or mention the other ways too. Since the man page covers the existence of ```USERNAME```, I think we could just leave the second sentence off here.

ryanlerch replied:
done.

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -494,8 +506,10 @@ def overrides():
 @click.option('--staging', help='Use the staging bodhi instance',
               is_flag=True, default=False)
+@click.option('--mine', is_flag=True,
+              help='Show only your overrides. User is defined by the $USERNAME envvar')
```
Same thing here - the user can be defined three ways. It might be best to just leave it at "Show only your overrides." so the helptext isn't too verbose. The USERNAME variable is defined in the man page.

bowlofeggs replied:
Optional: You could also use the man page for the expanded details about how the flag finds the username if you like.

ryanlerch replied:
removed the second sentence

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -301,9 +303,19 @@ def query(url, **kwargs):
     Args:
         url (unicode): The URL of a Bodhi server to create the update on. Ignored if staging is
                        True.
+        mine (Boolean): If the --mine flag was set
```
Let's add this and the other ```--mine``` flag to the man page at ```docs/man_pages/bodhi.rst```.

ryanlerch replied:
done. We were not documenting any options for overrides query, so i also added the --user option here too.

bowlofeggs commented on bodhi/tests/client/test_bindings.py:
```diff
@@ -88,8 +88,8 @@ def test_comment(self):
 
-class TestBodhiClient__init_username(unittest.TestCase):
-    """Test the BodhiClient._init_username() method."""
+class TestBodhiClientinit_username(unittest.TestCase):
```
Optional: might be nice to keep one _ between BodhiClient and init_username.

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -301,9 +303,19 @@ def query(url, **kwargs):
     """
     client = bindings.BodhiClient(base_url=url, staging=kwargs['staging'])
+    if mine:
+        if 'user' in kwargs:
```
I think this ```if``` statement might always be True. In my testing, I neither set the environment variable nor set the --user flag, but saw:

```
Warning: --mine and --user options both set. --user will be ignored.
```

In these cases, ```user``` is in ```kwargs``` but is set to ```None```. Perhaps the ```if``` statement here and elsewhere should be:

```
if kwargs.get('user', None):
```

bowlofeggs replied:
As I mentioned later, these warnings will also happen if the user uses the ```USERNAME``` environment variable, so perhaps we just don't need to warn here at all and can just drop the ```if``` statement?

ryanlerch replied:
done

bowlofeggs commented on bodhi/client/__init__.py:
```diff
@@ -508,11 +522,21 @@ def query_buildroot_overrides(url, user=None, **kwargs):
+    if mine:
+        if user:
+            click.echo('Warning: --mine and --user options both set. --user will be ignored.\n',
+                       err=True)
```
So one problem with this warning is that it will be displayed if the user uses ```USERNAME``` - that effectively comes in as ```--user```, so this causes the warning:

```
USERNAME="bowlofeggs" bodhi overrides query --mine
```

I wonder if it would be best to just not do these warnings? I struggle to think of a better way to handle it without falsely warning the user.

ryanlerch replied:
Dropped the warning!

bowlofeggs reviewed (approved):
Very nice!
